### PR TITLE
Safer load balancer reconfigurations

### PR DESCRIPTION
**Pull request by afzaledx**
Added a called to update_status() if an active server was found without IP address. If the IP is still missing after the call, allow the load balancer configuration to fail.

**JIRA tickets**: OC-4205

**Testing instructions**:

1. Spawn an appserver. but before making it active, terminate the OpenStack VM server. If we then try to make the appserver active. It should not be allowed to be made active. The VM server should be in the Ready state before making the appserver active.

2. When reconfiguring the loadbalancer, the VM server should have a public IP address, otherwise it will throw an exception. 

**Reviewers**
- [ ] @UmanShahzad

### CODE REVIEW

File: instance/models/openedx_instance.py [hunk 1]
```diff
@@ -117,6 +117,10 @@ def get_load_balancer_configuration(self, triggered_by_instance=False):
         for appserver in active_appservers:
             server_name = "appserver-{}".format(appserver.pk)
             ip_address = appserver.server.public_ip
+            if not ip_address:
+                # if the ip_address is not there, try to update_status() before failing.
```
Comment (UmanShahzad) [inline instance/models/openedx_instance.py]: This comment is a code smell: there is little that we gain from it which the code doesn't already make clear right after.

Also, by *fail*, we actually literally mean to make this function fail, i.e. cancel the load balancer reconfiguration process (This probably sounds confusing, so see https://github.com/open-craft/ansible-load-balancer#architecture-notes if it may be helpful), which probably entails throwing an exception or something to prevent Ocim from actually reconfiguring the load balancer in this state.

Comment (UmanShahzad) [reply instance/models/openedx_instance.py]: (BTW, for the comment itself, we can probably still have a comment but just rewrite it to explain in some more detail the motive behind this -- up to you).

File: instance/models/openedx_instance.py [hunk 2]
```diff
@@ -117,12 +117,19 @@ def get_load_balancer_configuration(self, triggered_by_instance=False):
...
+                appserver.server.update_status()
+                ip_address = appserver.server.public_ip
             if ip_address:
                 appserver_vars.append(dict(ip_address=ip_address, name=server_name))
             else:
                 appserver.logger.error(
                     "Active appserver does not have a public IP address. This should not happen."
+                    "Deconfiguring the load balancer backend."
                 )
+                return [], []
```
Comment (afzaledx) [inline instance/models/openedx_instance.py]: @UmanShahzad , this is pretty similar to what's happening on line 139. Raising an Exception would mess with the entire call stack.

File: instance/models/openedx_instance.py [hunk 3]
```diff
@@ -117,12 +117,21 @@ def get_load_balancer_configuration(self, triggered_by_instance=False):
...
             if ip_address:
                 appserver_vars.append(dict(ip_address=ip_address, name=server_name))
             else:
-                appserver.logger.error(
-                    "Active appserver does not have a public IP address. This should not happen."
+                self.logger.error(
+                    "Active appserver still does not have a public IP address after calling update_status()."
+                    "Deconfiguring the load balancer backend."
                 )
+                return [], []
```
Comment (UmanShahzad) [inline instance/models/openedx_instance.py]: This does nothing different than what it already would've, actually, for the case we're trying to prevent. Notice that `appserver_vars` will be empty if we don't return here if there is an issue retrieving the public IP, and it'll just go down to line 136, and do the same thing.

As mentioned earlier, we want to kill the call stack, because this case "should not happen" and if it does, it's too dangerous to proceed and potentially accidentally deconfigure an active and healthy backend.

File: instance/models/openedx_instance.py [hunk 4]
```diff
@@ -117,12 +117,21 @@ def get_load_balancer_configuration(self, triggered_by_instance=False):
         for appserver in active_appservers:
             server_name = "appserver-{}".format(appserver.pk)
             ip_address = appserver.server.public_ip
+            if not ip_address:
+                appserver.logger.error(
+                    "Active appserver does not have a public IP address. This should not happen."
+                    "Calling update_status()"
+                )
+                appserver.server.update_status()
+                ip_address = appserver.server.public_ip
```
Comment (UmanShahzad) [inline instance/models/openedx_instance.py]: :+1: Looks good.

If we introduce raising an exception instead of returning `[], []` in the case that the IP is still not available, that exception should probably go right here in this conditional block. Then later below we can erase the extra `if ip_address` check.

File: instance/models/openedx_instance.py [hunk 5]
```diff
@@ -117,12 +117,21 @@ def get_load_balancer_configuration(self, triggered_by_instance=False):
         for appserver in active_appservers:
             server_name = "appserver-{}".format(appserver.pk)
             ip_address = appserver.server.public_ip
+            if not ip_address:
+                appserver.logger.error(
+                    "Active appserver does not have a public IP address. This should not happen."
+                    "Calling update_status()"
```
Comment (UmanShahzad) [inline instance/models/openedx_instance.py]: We should probably not mention the function name explicitly, in case it has to change; then it'll have to change here, too, and additionally it's not as descriptive.

Better to just describe the process, i.e. "Updating internal status to double check.", or something along those lines.

File: instance/serializers/server.py [hunk 6]
```diff
@@ -50,8 +50,6 @@ class Meta:
         )
 
     def to_representation(self, obj):
-        # Make sure we have an up-to-date representation of the server
-        obj.update_status()
```
Comment (UmanShahzad) [inline instance/serializers/server.py]: :+1: looks good.

File: instance/tasks.py [hunk 7]
```diff
@@ -75,12 +76,15 @@ def spawn_appserver(
...
                 # If the AppServer provisioned successfully, make it active.
-                make_appserver_active(
-                    appserver_id,
-                    active=mark_active_on_success,
-                    deactivate_others=deactivate_old_appservers
-                )
-            break
+                appserver = OpenEdXAppServer.objects.get(pk=appserver_id)
+
+                if appserver.server.status == OpenStackServer.Status.Ready:
```
Comment (UmanShahzad) [inline instance/tasks.py]: This is the wrong level at which to check healthiness. It is *one* place, but we want to cover all possible places where activating an appserver takes place, and this is only one of them.

So we need to go to a lower level, in particular in the `make_appserver_active` function itself. It would be good to also log a note there for when the OpenStack server status is not correct for whatever reason.

Comment (UmanShahzad) [reply instance/tasks.py]: Also, I think we should import `Server` rather than `OpenStackServer` to get `.Status.Ready`, since it lets it be more generic for the future (i.e. an `AwsServer` would also be covered).

File: instance/api/openedx_appserver.py [hunk 8]
```diff
@@ -100,7 +100,12 @@ def make_active(self, request, pk):
             return Response(
                 {"error": "Cannot make an unhealthy app server active."}, status=status.HTTP_400_BAD_REQUEST
             )
-        make_appserver_active(app_server.pk)
+        successfully_activated = make_appserver_active(app_server.pk)
```
Comment (UmanShahzad) [inline instance/api/openedx_appserver.py]: Similar to my comment earlier, we can move this into the previous `if` check to decrease repeat code, i.e.

```python
if not app_server.status.is_healthy_state or not make_appserver_active(app_server.pk):
         return Response(
             {"error": "Cannot make an unhealthy app server active."}, status=status.HTTP_400_BAD_REQUEST
         )
```

This'll make it active if the appserver is healthy and move on if successful, otherwise fail with the same error.

File: instance/models/openedx_instance.py [hunk 9]
```diff
@@ -117,12 +118,17 @@ def get_load_balancer_configuration(self, triggered_by_instance=False):
...
-                appserver_vars.append(dict(ip_address=ip_address, name=server_name))
-            else:
+            if not ip_address:
                 appserver.logger.error(
                     "Active appserver does not have a public IP address. This should not happen."
+                    "Updating internal status to double check."
                 )
+                appserver.server.update_status()
+                ip_address = appserver.server.public_ip
+                if not ip_address:
```
Comment (UmanShahzad) [inline instance/models/openedx_instance.py]: Nit: since we're not reusing `ip_address` now, can we just do `if not appserver.server.public_ip`?

File: instance/models/openedx_instance.py [hunk 10]
```diff
@@ -117,12 +118,17 @@ def get_load_balancer_configuration(self, triggered_by_instance=False):
...
+                    raise WrongStateException("Active appserver does not have a public IP address.")
```
Comment (UmanShahzad) [inline instance/models/openedx_instance.py]: Nit: can we improve this a bit to tell a story? :D I.e., "Public IP still not available for active appserver. Canceling reconfiguration process."

File: instance/models/openedx_instance.py [hunk 11]
```diff
@@ -117,12 +118,17 @@ def get_load_balancer_configuration(self, triggered_by_instance=False):
         for appserver in active_appservers:
             server_name = "appserver-{}".format(appserver.pk)
             ip_address = appserver.server.public_ip
-            if ip_address:
-                appserver_vars.append(dict(ip_address=ip_address, name=server_name))
-            else:
+            if not ip_address:
                 appserver.logger.error(
                     "Active appserver does not have a public IP address. This should not happen."
+                    "Updating internal status to double check."
```
Comment (UmanShahzad) [inline instance/models/openedx_instance.py]: We either need a space at the end of the previous line, or in front of this one. Otherwise they'll be concatenated awkwardly.

File: instance/tasks.py [hunk 12]
```diff
@@ -93,15 +95,22 @@ def make_appserver_active(appserver_id, active=True, deactivate_others=False):
...
+    if appserver.server.status == Server.Status.Ready:
+        appserver.make_active(active=active)
+        success = True
+        if active and deactivate_others:
+            other_appservers = appserver.instance.appserver_set.filter(_is_active=True).exclude(pk=appserver_id)
+            for appserver_to_deactivate in other_appservers:
+                logger.info('Deactivating %s [%s]', appserver_to_deactivate, appserver_to_deactivate.id)
+                appserver_to_deactivate.make_active(active=False)
+    else:
+        success = False
```
[the remark this answers is not in the document]
Comment (afzaledx) [reply instance/tasks.py]: @UmanShahzad, what if `active` was passed in as `False`, i.e. when deactivating the app server? It does not seem appropriate to not deactivate when the appserver.server is not ready.

[the remark this answers is not in the document]
Comment (UmanShahzad) [reply instance/tasks.py]: With the code as you currently have it, this will still happen -- deactivating the appserver only happens under the condition of the server status being ready.

The only case I can think of right now where the VM is not 'Ready' but is still active is in the 'Unknown' state, and in that state we don't want to deactivate anyway right, because it's dangerous?

[the remark this answers is not in the document]
Comment (afzaledx) [reply instance/tasks.py]: @UmanShahzad , I have changed the logic around a bit. can you take a look at it again?

[the remark this answers is not in the document]
Comment (UmanShahzad) [reply instance/tasks.py]: @afzaledx I took a look, and it will do as you said, but is it what we want? See my last paragraph above.

> The only case I can think of right now where the VM is not 'Ready' but is still active is in the 'Unknown' state, and in that state we don't want to deactivate anyway right, because it's dangerous?

File: instance/tasks.py [hunk 13]
```diff
@@ -92,16 +91,28 @@ def spawn_appserver(
...
     appserver = OpenEdXAppServer.objects.get(pk=appserver_id)
-    logger.info('%s %s: ID=%s', "Activating" if active else "Deactivating", appserver, appserver_id)
-    appserver.make_active(active=active)
-    if active and deactivate_others:
-        other_appservers = appserver.instance.appserver_set.filter(_is_active=True).exclude(pk=appserver_id)
-        for appserver_to_deactivate in other_appservers:
-            logger.info('Deactivating %s [%s]', appserver_to_deactivate, appserver_to_deactivate.id)
-            appserver_to_deactivate.make_active(active=False)
+    vmserver_ready = appserver.server.status == Server.Status.Ready
+
```
Comment (afzaledx) [inline instance/tasks.py]: @UmanShahzad , this is the new code. you are looking at this ? I am a bit confused. We do want to deactivate without checking the vm server, right?

File: instance/tasks.py [hunk 14]
```diff
@@ -92,16 +91,28 @@ def spawn_appserver(
 def make_appserver_active(appserver_id, active=True, deactivate_others=False):
     """
     Mark an AppServer as active or inactive.
+
+    :returns True if successful, False otherwise.
```
Comment (UmanShahzad) [inline instance/tasks.py]: I would clarify this comment a bit more to be explicit about what successful really means in this function's context.

File: instance/tasks.py [hunk 15]
```diff
@@ -92,16 +91,28 @@ def spawn_appserver(
...
     """
     appserver = OpenEdXAppServer.objects.get(pk=appserver_id)
-    logger.info('%s %s: ID=%s', "Activating" if active else "Deactivating", appserver, appserver_id)
-    appserver.make_active(active=active)
-    if active and deactivate_others:
-        other_appservers = appserver.instance.appserver_set.filter(_is_active=True).exclude(pk=appserver_id)
-        for appserver_to_deactivate in other_appservers:
-            logger.info('Deactivating %s [%s]', appserver_to_deactivate, appserver_to_deactivate.id)
-            appserver_to_deactivate.make_active(active=False)
+    vmserver_ready = appserver.server.status == Server.Status.Ready
```
Comment (UmanShahzad) [inline instance/tasks.py]: A nit that came up: maybe we should move this check to be a property of `appserver.server`? I.e. `appserver.server.is_healthy_state` or something.

File: instance/tasks.py [hunk 16]
```diff
@@ -92,16 +91,28 @@ def spawn_appserver(
...
+        if vmserver_ready:
+            logger.info('Activating %s: ID=%s', appserver, appserver_id)
+            appserver.make_active(active=True)
+            if deactivate_others:
+                other_appservers = appserver.instance.appserver_set.filter(_is_active=True).exclude(pk=appserver_id)
+                for appserver_to_deactivate in other_appservers:
+                    logger.info('Deactivating %s [%s]', appserver_to_deactivate, appserver_to_deactivate.id)
+                    appserver_to_deactivate.make_active(active=False)
+        else:
+            logger.info('Did not Activate %s: ID=%s. The server was not Ready.', appserver, appserver_id)
```
Comment (UmanShahzad) [inline instance/tasks.py]: @afzaledx Btw, one more note for when you change this, I _think_ we should do something similar here as above where since the VM is unhealthy, we re-check the status with appserver.server.update_status, and check again if it is healthy, in case the status was 'Unknown' and we need to just refresh it with another call to the Nova API.

File: instance/tasks.py [hunk 17]
```diff
@@ -75,12 +76,13 @@ def spawn_appserver(
                 instance.tags.add(success_tag)
```
[the remark this answers is not in the document]
Comment (afzaledx) [reply instance/tasks.py]: @UmanShahzad , Still trying to figure out how to do that in a proper way, since tags is a 1 to many field, not a simple list.

File: instance/models/server.py [hunk 18]
```diff
@@ -70,6 +70,10 @@ class ServerState(ResourceState):
     # until the target server has reached one of these statuses.
     vm_available = False
 
+    # We want to reconfigure the load balancer in case the server is in Status.Ready.
+    # This field be set just in the Ready state and False in other states.
```
Comment (UmanShahzad) [inline instance/models/server.py]: "This field *is to* be set...".

File: instance/models/server.py [hunk 19]
```diff
@@ -70,6 +70,10 @@ class ServerState(ResourceState):
     # until the target server has reached one of these statuses.
     vm_available = False
 
+    # We want to reconfigure the load balancer in case the server is in Status.Ready.
```
Comment (UmanShahzad) [inline instance/models/server.py]: "We *can only* reconfigure the load balancer...".

File: instance/tests/test_tasks.py [hunk 20]
```diff
@@ -146,13 +172,29 @@ def test_make_appserver_active(self, active, mock_make_active):
...
+                mock_make_active.assert_called_once_with(active=True)
+
     @patch('instance.models.openedx_appserver.OpenEdXAppServer.make_active')
     def test_deactivate_others(self, mock_make_active):
         """
         When activating the appserver, optionally deactivate others.
         """
         for dummy in range(5):
             appserver = make_test_appserver(self.appserver.instance)
+            appserver.server = ReadyOpenStackServerFactory()
```
Comment (UmanShahzad) [inline instance/tests/test_tasks.py]: I noticed we are adding this in **lot** of different places. This is bad: it's code repetition everywhere. Please move this to a common place, like `make_test_appserver`, so it's not done like this. You can add it as a keyword arg that'll default to the ready OpenStack server, or whichever way you think is best.

File: instance/tests/models/test_openedx_instance.py [hunk 21]
```diff
@@ -457,7 +459,7 @@ def test_get_load_balancer_configuration(self, mocks):
                 self.assertLogs("instance.models.instance", "ERROR"):
             appserver.server._public_ip = None
             appserver.server.save()
-            self.assertEqual(instance.get_load_balancer_configuration(), ([], []))
+            self.assertRaises(WrongStateException, instance.get_load_balancer_configuration)
```
Comment (UmanShahzad) [inline instance/tests/models/test_openedx_instance.py]: Please change the comment above this `with` block to reflect the new reality of what happens.

File: instance/tests/models/test_openedx_instance.py [hunk 22]
```diff
@@ -733,9 +735,10 @@ def test_archive_no_active_appserver(self, mock_reconfigure, mock_disable_monito
         ])
         self.assertTrue(instance.ref.is_archived)
 
+    @patch('instance.models.server.OpenStackServer.public_ip')
```
Comment (UmanShahzad) [inline instance/tests/models/test_openedx_instance.py]: I don't understand immediately -- why is this needed now?

File: instance/tests/test_tasks.py [hunk 23]
```diff
@@ -72,29 +73,53 @@ def test_mark_active_on_success(self, provisioning_succeeds):
         Test that we when mark_active_on_success=True, the spawn_appserver task will mark the
```
Comment (UmanShahzad) [inline instance/tests/test_tasks.py]: I know you didn't comment this, but since we're here can you remove that extraneous 'we'? Thanks :P

File: instance/tests/test_tasks.py [hunk 24]
```diff
@@ -146,13 +172,29 @@ def test_make_appserver_active(self, active, mock_make_active):
         tasks.make_appserver_active(self.appserver.id, active=active)
         mock_make_active.assert_called_once_with(active=active)
 
+    @ddt.data([False, True], [False, False])
+    @patch('instance.models.server.OpenStackServer.update_status')
+    @patch('instance.models.openedx_appserver.OpenEdXAppServer.make_active')
+    def test_make_appserver_active_server_unhealthy(self, health, mock_make_active, mock_update_status):
```
Comment (UmanShahzad) [inline instance/tests/test_tasks.py]: Can we call this `test_make_appserver_active_but_server_unhealthy`?

File: instance/tests/test_tasks.py [hunk 25]
```diff
@@ -146,13 +172,29 @@ def test_make_appserver_active(self, active, mock_make_active):
...
+        """
+        By default, we activate the appserver.
+        """
+        with patch('instance.models.server.OpenStackServer.Status.Ready.is_healthy_state',
+                   new_callable=PropertyMock) as mock_is_healthy_state:
+            mock_is_healthy_state.side_effect = health
+            tasks.make_appserver_active(self.appserver.id, active=True)
+            mock_update_status.assert_called_once_with()
+            if not all(not data for data in health):
```
Comment (UmanShahzad) [inline instance/tests/test_tasks.py]: This is hard to read. It's a `not` operation applied on a group of `AND`'d `not`s, so it expands to be a group of just `or`s. So this can be simplified to `if any(data for data in health)`. (I'd rename `data` to `healthy`, btw).

(For an example, see this:)
```
~(~a and ~b and ~c)
= a or b or c
```

And in the load-balancer reconfiguration context, if we see that the server is healthy _any_ time, it's good to be made active.

File: instance/tests/models/test_openedx_instance.py (hunk 22 above)
Comment (afzaledx) [reply instance/tests/models/test_openedx_instance.py]: the call to public_ip is an external call. On line 772, there's a call to
        # Set single app server active - this will be the reference date
        with freeze_time(reference_date):
            active_appserver.make_active()

so that will check the public_ip. Need to patch the public_ip call to avoid failures due to the external API call. Basically whereever the make_active() funciton is called, we need to patch the public_ip because we have added it as a precondition before making the appserver active.

File: instance/tests/test_tasks.py (hunk 25 above)
Comment (afzaledx) [reply instance/tests/test_tasks.py]: Thanks for that. Really helpful.

File: instance/tests/models/test_openedx_instance.py (hunk 22 above)
Comment (afzaledx) [reply instance/tests/models/test_openedx_instance.py]: Well, I removed the mock and it passed anyway, so removing it now.

File: instance/tests/models/factories/openedx_appserver.py [hunk 26]
```diff
@@ -37,4 +37,10 @@ def make_test_appserver(instance=None):
     if not instance.load_balancing_server:
         instance.load_balancing_server = LoadBalancingServer.objects.select_random()
         instance.save()
-    return instance._create_owned_appserver()
+    appserver = instance._create_owned_appserver()
+
+    if server:
```
Comment (UmanShahzad) [inline instance/tests/models/factories/openedx_appserver.py]: This doesn't fix the problem of repeat code I mentioned -- we're still just repeating the creation of `ReadyOpenStackServerFactory` everywhere.

What we should do instead is, `appserver.server = ReadyOpenStackServerFactory() if server is None else server`.

File: instance/tests/models/test_openedx_instance.py (hunk 22 above)
Comment (UmanShahzad) [reply instance/tests/models/test_openedx_instance.py]: :+1:

File: instance/tests/test_tasks.py [hunk 27]
```diff
@@ -146,6 +170,22 @@ def test_make_appserver_active(self, active, mock_make_active):
         tasks.make_appserver_active(self.appserver.id, active=active)
         mock_make_active.assert_called_once_with(active=active)
 
+    @ddt.data([False, True], [False, False])
+    @patch('instance.models.server.OpenStackServer.update_status')
+    @patch('instance.models.openedx_appserver.OpenEdXAppServer.make_active')
+    # pylint: disable=invalid-name
```
Comment (UmanShahzad) [inline instance/tests/test_tasks.py]: Ah snap, I didn't know this would introduce a pylint violation :) In that case, sorry for my old comment -- it's probably better to leave this as you originally had it.

File: instance/tests/models/factories/openedx_appserver.py (hunk 26 above)
Comment (afzaledx) [reply instance/tests/models/factories/openedx_appserver.py]: Well, I changed the default to Ready state, and 20 tests started failing, because the default server being created is in the Pending state, and that's why we have to overwrite it with the Ready state server. I think it's suitable to use it they what I have it now because it's very similar to how `instance` is being used.

Comment (UmanShahzad) [reply instance/tests/models/factories/openedx_appserver.py]: Ah, OK, I did not know about that default. Sorry, my recommendation was not very well informed here. But the addition to the make_test_appserver function is still pretty nice, thanks!

File: instance/tests/models/test_openedx_instance.py (hunk 22 above)
Comment (afzaledx) [reply instance/tests/models/test_openedx_instance.py]: actually it didn't pass. had to add it back. Sorry about the confusion.

**Comment by UmanShahzad** [comment]
@afzaledx Can you squash commits here and get these tests to pass? Then we should be good for merge!
